migrations: populate initial version setting

tbg opened the pull request:
Add a bootstrap version which is populated when a cluster running >1.0 is
bootstrapped. Use it to populate the settings table via a sql migration. At the
same time, prevent `SET CLUSTER SETTING version = 'x'` from working until that
migration has run.

This solves one remaining headache for [version migrations][1] by giving it
authoritative information on the currently running cluster version during
upgrades.

Fixes #17389.

[1]: https://github.com/cockroachdb/cockroach/pull/16977

tbg committed:
storage: persist version on cluster bootstrap

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17694)
<!-- Reviewable:end -->

tbg committed:
migrations: populate initial version setting

This uses the previously added bootstrap version which is populated when a
cluster running >1.0 is bootstrapped to populate the settings table. At the
same time, prevent `SET CLUSTER SETTING version = 'x'` from working until
that migration has run.

This solves one remaining headache for [version migrations][1] by giving it
authoritative information on the currently running cluster version during
upgrades.

Fixes #17389.

[1]: https://github.com/cockroachdb/cockroach/pull/16977

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 8 of 8 files at r1, 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/migration/sqlmigrations/migrations.go, line 416 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17694#-KrgyLZ3eJ9Bd3FAlgPt:-KrgyLZ4baVFQuurI3Hj:bwqh2n7) ([raw file](https://github.com/cockroachdb/cockroach/blob/eddc62caa7285b896304b56fd2a55b070d70671d/pkg/migration/sqlmigrations/migrations.go#L416)):*
> ```Go
> func populateVersionSetting(ctx context.Context, r runner) error {
> 	var v roachpb.Version
> 	if err := r.db.Txn(ctx, func(ctx context.Context, txn *client.Txn) error {
> ```

Can this be `r.db.Get` instead of using an explicit Txn?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17694#-:-Krgyl9CY4d4wJ4Kun1r:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/migration/sqlmigrations/migrations.go, line 416 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17694#-KrgyLZ3eJ9Bd3FAlgPt:-KrgzPHx4Ufh5mPxNaeP:b-j8jyxe) ([raw file](https://github.com/cockroachdb/cockroach/blob/eddc62caa7285b896304b56fd2a55b070d70671d/pkg/migration/sqlmigrations/migrations.go#L416)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Can this be `r.db.Get` instead of using an explicit Txn?
</blockquote></details>

Yes, but I'd have to plumb more stuff onto `runner`, and so I decided against it. I can add a comment and update the runner, whichever you prefer.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17694)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/migration/sqlmigrations/migrations.go, line 416 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17694#-KrgyLZ3eJ9Bd3FAlgPt:-Krh-J_VRXMk5E-oHbgF:bhrgrr3) ([raw file](https://github.com/cockroachdb/cockroach/blob/eddc62caa7285b896304b56fd2a55b070d70671d/pkg/migration/sqlmigrations/migrations.go#L416)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Yes, but I'd have to plumb more stuff onto `runner`, and so I decided against it. I can add a comment and update the runner, whichever you prefer.
</blockquote></details>

Leaving it is fine; I didn't realize this wasn't a full `client.DB`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17694)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTR, Ben!
